server/sonar-server/src/main/java/org/sonar/server/component/index/ComponentIndex.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.component.index;
+
+import java.util.List;
+import org.elasticsearch.action.search.SearchRequestBuilder;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.sonar.server.es.BaseIndex;
+import org.sonar.server.es.EsClient;
+import org.sonar.server.es.SearchIdResult;
+
+import static org.elasticsearch.index.query.QueryBuilders.boolQuery;
+import static org.elasticsearch.index.query.QueryBuilders.matchQuery;
+import static org.elasticsearch.index.query.QueryBuilders.termQuery;
+import static org.sonar.server.component.index.ComponentIndexDefinition.FIELD_KEY;
+import static org.sonar.server.component.index.ComponentIndexDefinition.FIELD_NAME;
+import static org.sonar.server.component.index.ComponentIndexDefinition.FIELD_QUALIFIER;
+import static org.sonar.server.component.index.ComponentIndexDefinition.INDEX_COMPONENTS;
+import static org.sonar.server.component.index.ComponentIndexDefinition.TYPE_COMPONENT;
+
+public class ComponentIndex extends BaseIndex {
+
+  public ComponentIndex(EsClient client) {
+    super(client);
+  }
+
+  public List<String> search(ComponentIndexQuery query) {
+    SearchRequestBuilder requestBuilder = getClient()
+      .prepareSearch(INDEX_COMPONENTS)
+      .setTypes(TYPE_COMPONENT)
+      .setFetchSource(false);
+
+    query.getLimit().ifPresent(requestBuilder::setSize);
+
+    requestBuilder.setQuery(
+      createQuery(query));
+
+    return new SearchIdResult<>(requestBuilder.get(), id -> id).getIds();
+  }
+
+  private static QueryBuilder createQuery(ComponentIndexQuery query) {
+
+    BoolQueryBuilder esQuery = boolQuery();
+
+    if (query.hasQualifiers()) {
+      BoolQueryBuilder qualifierQuery = boolQuery();
+
+      query
+        .getQualifiers()
+        .stream()
+        .map(qualifier -> termQuery(FIELD_QUALIFIER, qualifier))
+        .forEach(qualifierQuery::should);
+
+      esQuery.filter(qualifierQuery.minimumNumberShouldMatch(1));
+    }
+
+    return esQuery
+      .filter(
```

Comment:
Because they contribute to score, criteria on name and key must be queries but not filters (`esQuery.should(...)`).